File: libraries/php/src/universal_utilities/SimpleTimer.php
<?php
/**
 * Created by PhpStorm.
 * User: utarsuno
 * Date: 2019-03-24
 * Time: 15:39
 */

namespace QuasarSource\Utilities;


class SimpleTimer {

    private $running;
    private $time_start;
    private $time_end;
    private $time_accumulated;

    public function __construct(bool $auto_start=false) {
        $this->running          = false;
        $this->time_start       = $auto_start ? microtime(true) : null;
        $this->time_accumulated = 0.0;
    }

    public function start() : void {
        if (!$this->running) {
            $this->time_start = microtime(true);
            $this->running    = true;
        }
    }

    public function stop() : void {
        if ($this->running) {
            $this->time_end = microtime(true);
            $this->running  = false;
        }
    }

    public function pause() : void {
        if ($this->running) {
            if ($this->time_end === null) {
                $this->stop();
            }
            $this->time_accumulated += $this->time_end - $this->time_start;
            $this->running           = false;
        }
    }

    public function __toString() : string {
        $this->pause();
        if ($this->time_accumulated < 0.0001) {
            return '0s';
        }
        return strval($this->time_accumulated) . 's';
    }

}

File: libraries/python/universal_utilities/time_abstraction/simple_timer.py
# coding=utf-8

"""This module, simple_timer.py, provides a class for a simple timer."""

import time


class SimpleTimer(object):
	"""A simple timer."""

	def __init__(self, auto_start=False):
		self._running          = False
		self._time_start       = None
		self._time_end         = None
		self._time_accumulated = 0.0
		if auto_start:
			self.start()

	def start(self):
		"""Starts the timer."""
		assert not self._running
		self._time_start = time.time()
		self._running    = True

	def stop(self):
		"""Ends the timer."""
		assert self._running
		self._time_end = time.time()
		self._running  = False

	def pause(self):
		"""Pause the timer."""
		if self._running:
			if self._time_end is None:
				self.stop()
			self._time_accumulated += self._time_end - self._time_start
			self._running           = False

	def __str__(self):
		self.pause()
		if self._time_accumulated < 0.0001:
			return '0s'
		return str(self._time_accumulated) + 's'



